src/interrupt.rs

```diff
@@ -30,42 +30,127 @@
 
 pub use bare_metal::{CriticalSection, Mutex, Nr};
 
+use core::marker::PhantomData;
 #[cfg(target_arch = "avr")]
 use core::arch::asm;
 
-#[inline]
-/// Disables all interrupts
+/// Opaque structure for storing the global interrupt flag status.
 ///
-/// Returns a bool, reflecting whether interrupts were enabled prior to calling this method.
-pub fn disable() -> bool {
+/// This structure does not implement `Copy` and `Clone`,
+/// because the user shall not duplicate and pass it twice to [crate::interrupt::restore].
+#[derive(Debug)]
+pub struct IrqFlag {
+    // The saved SREG.
+    sreg: u8,
+    // Use PhantomData with raw pointer to disable Send and Sync auto-trait impls.
+    // Negative impl of Send/Sync is not stabilized in rustc, yet.
+    // This is not strictly necessary for memory safety.
+    // It's just to keep the user from passing this struct between interrupt contexts,
+    // which would not be meaningful.
+    _phantom: PhantomData<*mut u8>,
+}
+
+impl IrqFlag {
+    #[inline(always)]
+    fn new(sreg: u8) -> IrqFlag {
+        IrqFlag {
+            sreg,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Check the status of the saved global interrupt flag.
+    ///
+    /// Returns true, if the saved global interrupt flag is set (IRQs enabled).
+    /// Otherwise returns false.
+    ///
+    /// This method can be used to check whether interrupts were enabled
+    /// before the [crate::interrupt::disable_save] call.
+    /// You probably shouldn't make your program behavior dependent on this state.
+    /// Consider using a different design.
+    #[inline(always)]
+    pub fn enabled(&self) -> bool {
+        self.sreg & 0x80 != 0
+    }
+}
+
+// PhantomData is not supported by uDebug. Spin our own fmt implementation.
+impl ufmt::uDebug for IrqFlag {
+    fn fmt<W: ufmt::uWrite + ?Sized>(&self, f: &mut ufmt::Formatter<'_, W>) -> Result<(), W::Error> {
+        f.debug_struct("IrqFlag")?
+            .field("sreg", &self.sreg)?
+            .finish()
+    }
+}
```

Comment:
Of course, when making `ufmt` optional, this also needs to be gated behind a `#[cfg(feature = "ufmt")]`

Reply:
Has been replaced by an optional #derive.